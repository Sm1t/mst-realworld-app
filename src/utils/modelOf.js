import { isStateTreeNode } from 'mobx-state-tree';

export default type => {
  const fn = (props, propName) => {
    if (isStateTreeNode(props[propName]) && type.is(props[propName])) {
      return null;
    }
    return new Error(`Invalid value for prop ${propName}`);
  };

  fn.isRequired = (props, propName) => !(propName in props)
    ? new Error(`Missing prop ${propName} in props`)
    : fn(props, propName);

  return fn;
};
