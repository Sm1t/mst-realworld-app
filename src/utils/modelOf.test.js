import { types } from 'mobx-state-tree';

import modelOf from './modelOf';

const model = types.model({});
const store = model.create();

describe('modelOf', () => {
  it('should return null when prop type is valid', () => {
    const result = modelOf(model)({ store }, 'store');

    expect(result).toBeNull();
  });

  it('should return properly error when value for prop is invalid', () => {
    const result = modelOf(model)({ store: {} }, 'store');

    expect(result).toBeInstanceOf(Error);
    expect(result.message).toMatchSnapshot();
  });

  describe('with isRequired', () => {
    it('should return properly error when missing required prop', () => {
      const result = modelOf(model).isRequired({}, 'store');

      expect(result).toBeInstanceOf(Error);
      expect(result.message).toMatchSnapshot();
    });

    it('should properly validate required prop', () => {
      const result = modelOf(model).isRequired({ store }, 'store');

      expect(result).toBeNull();
    });
  });
});
