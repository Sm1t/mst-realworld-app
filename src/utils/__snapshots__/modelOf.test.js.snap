// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modelOf should return properly error when value for prop is invalid 1`] = `"Invalid value for prop store"`;

exports[`modelOf with isRequired should return properly error when missing required prop 1`] = `"Missing prop store in props"`;
